tests/models/test_cnnlstm.py

```diff
@@ -0,0 +1,856 @@
+        assert output['probs'].shape == (2, 10)
+
+    def test_cnnlstm_forward_with_lengths(
+        self, compatible_components: dict[str, Any]
+    ) -> None:
+        """Test CNNLSTM forward with sequence lengths."""
+        model = CNNLSTM(**compatible_components)
+        x = torch.randn(3, 6, 3, 224, 224)
+        lengths = torch.tensor([6, 4, 2])
+        output = model(x, lengths=lengths)
+
+        assert output['outputs'].shape == (3, 10)
+        assert output['probs'].shape == (3, 10)
+
+    def test_cnnlstm_forward_with_kwargs(
+        self, compatible_components: dict[str, Any]
+    ) -> None:
+        """Test CNNLSTM forward accepts additional kwargs."""
+        model = CNNLSTM(**compatible_components)
+        x = torch.randn(1, 3, 224, 224)
+        output = model(x, dummy_kwarg='test')  # Should not fail
+        assert output['outputs'].shape == (1, 10)
+
+    def test_cnnlstm_regression_task(self) -> None:
+        """Test CNNLSTM with regression task."""
+        cnn = CNN(model_name='resnet18', pretrained=False, output_dim=128)
+        lstm = LSTM(input_dim=128, hidden_dim=64, bidirectional=True)  # output: 128
+        task = Task(input_dim=128, task_type='regression', output_dim=3)
+
+        model = CNNLSTM(cnn_backbone=cnn, rnn_encoder=lstm, task_head=task)
+        x = torch.randn(2, 4, 3, 224, 224)
+        output = model(x)
+
+        assert output['outputs'].shape == (2, 3)
+        assert output['probs'] is None
+
+    def test_cnnlstm_bidirectional_lstm(self) -> None:
+        """Test CNNLSTM with bidirectional LSTM."""
+        cnn = CNN(model_name='resnet18', pretrained=False, output_dim=128)
+        lstm = LSTM(input_dim=128, hidden_dim=64, bidirectional=True, num_layers=2)
+        task = Task(
+            input_dim=128, task_type='classification', num_classes=5, pooling='mean'
+        )
+
+        model = CNNLSTM(cnn_backbone=cnn, rnn_encoder=lstm, task_head=task)
+        x = torch.randn(2, 5, 3, 224, 224)
+        output = model(x)
+
+        assert output['outputs'].shape == (2, 5)
+        assert output['probs'].shape == (2, 5)
+
+    def test_cnnlstm_gradient_flow(self, compatible_components: dict[str, Any]) -> None:
+        """Test gradient flow through CNNLSTM."""
+        model = CNNLSTM(**compatible_components)
+        x = torch.randn(1, 3, 3, 224, 224, requires_grad=True)
+
+        output = model(x)
+        loss = output['outputs'].sum()
+        loss.backward()
+
+        # Check input gradients
+        assert x.grad is not None
+        assert x.grad.shape == x.shape
+
+        # Check parameter gradients
+        params_with_grad = 0
+        total_params = 0
+        for param in model.parameters():
+            if param.requires_grad:
+                total_params += 1
+                if param.grad is not None:
+                    params_with_grad += 1
+
+        assert params_with_grad > 0, 'No parameter gradients found'
+
+
+class TestIntegrationAndEdgeCases:
```

Comment:
Let's merge the rest of these classes into one of the above. Running `pytest tests/models/test_cnnlstm.py::TestCNN` should result in 100% coverage for CNN. Any integration tests can go in `TestCNNLSTM` since it represents the total sum of all models.